chainercv/evaluations/eval_semantic_segmentation_iou.py

```diff
@@ -0,0 +1,139 @@
+from __future__ import division
+
+import numpy as np
+import six
+
+
+def calc_semantic_segmentation_confusion(pred_labels, gt_labels, n_class):
+    """Collect confusion matrix.
+
+    Args:
+        pred_labels (iterable of numpy.ndarray): A collection of predicted
+            labels. The shape of a label array
+            is :math:`(H, W)`. :math:`H` and :math:`W`
+            are height and width of the label.
+        gt_labels (iterable of numpy.ndarray): A collection of ground
+            truth label. The shape of a ground truth label array is
+            :math:`(H, W)`. The corresponding prediction label should
+            have the same shape.
+            A pixel with value :obj:`-1` will be ignored during evaluation.
+        n_class (int): The number of classes.
+
+    Returns:
+        numpy.ndarray:
+        A confusion matrix. Its shape is :math:`(n\_class, n\_class)`.
+        The :math:`(i, j)` th element corresponds to the number of pixels
+        that are labeled as class :math:`i` by the ground truth and
+        class :math:`j` by the prediction.
+
+    """
+    pred_labels = iter(pred_labels)
+    gt_labels = iter(gt_labels)
+
+    confusion = np.zeros((n_class, n_class), dtype=np.int64)
+    for pred_label, gt_label in six.moves.zip(pred_labels, gt_labels):
+        if pred_label.ndim != 2 or gt_label.ndim != 2:
+            raise ValueError('ndim of inputs should be two.')
+        if pred_label.shape != gt_label.shape:
+            raise ValueError('Shapes of inputs should be same.')
+
+        pred_label = pred_label.flatten()
+        gt_label = gt_label.flatten()
+        mask = (gt_label >= 0) & (gt_label < n_class)
+        confusion += np.bincount(
+            n_class * gt_label[mask].astype(int) +
+            pred_label[mask], minlength=n_class**2).reshape(n_class, n_class)
```

Comment:
I think we can expand the confusion matrix dynamically. `n_class` can be removed.
```
n_class = 0
...
lb_max = gt_label.max()
if  lb >= n_class:
    expanded_confusion = np.zeros((lb + 1, lb + 1))
    expanded_confusion[0:n_class, 0:n_class] = confusion

    n_class = lb + 1
    confusion = expanded_confusion
```
This expansion may affect performance. But I think it will be ignorable because neural networks are much slower.
How do you think of it?